Realm/Realm/Sync/ErrorHandling/RecoverOrDiscardUnsyncedChangesHandler.cs

```diff
@@ -0,0 +1,88 @@
+    /// In this situation this strategy fallsback to the discard unsynced one.
+    /// To reiterate what it does: it discards all the unsynced local changes and uses the latest realm that is available on the remote sync server.
+    /// You can read more about the automatic merge rules at <see href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client Resets - .NET SDK</see>.
+    /// The automatic recovery mechanism creates write transactions meaning that all the changes that take place
+    /// are properly propagated through the standard Realm's change notifications.
+    /// The <see cref="RecoverOrDiscardUnsyncedChangesHandler"/> strategy supplies four callbacks: <see cref="OnBeforeReset"/>, <see cref="OnAfterRecovery"/>,
+    /// <see cref="OnAfterDiscard"/> and <see cref="ManualResetFallback"/>.
+    /// The first callback is invoked just before the client reset happens. While the second and third callbacks are respectively called after an automatic
+    /// client reset succeeded, or the automatic client reset failed and instead the discard unsynced one succeded.
+    /// The last callback is invoked whenever an error occurs in either of the recovery stragegies and the system needs to fallback to a manual mode.
+    /// The overall recommendation for using this strategy is that using the three available callbacks should only be considered when:
+    /// 1. The user needs to be notified (in <see cref="OnBeforeReset"/>) of an incoming potential data loss
+    ///    of unsynced data as a result of a merge or a complete discard of unsynced local changes
+    /// 2. The user needs to be notified (in <see cref="OnAfterRecovery"/> or <see cref="OnAfterDiscard"/>)
+    ///    that the reset process has completed
+    /// 3. Advanced use cases for data-sensitive applications where the developer wants
+    ///    to recover in the most appropriate way the unsynced data
+    /// 4. Backup the whole realm before the client reset happens (in <see cref="OnBeforeReset"/>).
+    ///    Such backup could, for example, be used to restore the unsynced data (see 3.)
+    /// </remarks>
+    /// <seealso href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client Resets - .NET SDK</seealso>
+    public sealed class RecoverOrDiscardUnsyncedChangesHandler : ClientResetHandlerBase
+    {
+        internal override ClientResyncMode ClientResetMode => ClientResyncMode.RecoverOrDiscard;
+
+        /// <summary>
+        /// Gets or sets the callback that indicates a Client Reset is about to happen.
+        /// </summary>
+        /// <value>Callback invoked right before a Client Reset.</value>
+        public BeforeResetCallback OnBeforeReset { get; set; }
+
+        /// <summary>
+        /// Gets or sets the callback that indicates that an automatic Client Reset just happened.
+        /// </summary>
+        /// <value>Callback invoked right after a Client Reset.</value>
```

Comment:
In the remarks we should indicate that only one of `OnAfterRecovery` or `OnAfterDiscard` but not both.